src/main/java/com/hackathonorganizer/hackathonwriteservice/team/service/TeamService.java

```diff
@@ -0,0 +1,93 @@
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class TeamService {
+
+
+    private final TeamRepository teamRepository;
+    private final TagService tagService;
+    //TODO change to service when it will be ready
+    private final HackathonRepository hackathonRepository;
+
+    public TeamResponse create(TeamRequest teamRequest) {
+        val teamToSave = Team.builder()
+                .ownerId(teamRequest.ownerId())
+                .hackathon(getHackathon(teamRequest.hackathonId()))
+                .teamMembersIds(teamRequest.teamMembersIds())
+                .tags(getTags(teamRequest.tags()))
+                .build();
+        log.info("Trying to save new team {}", teamToSave);
+        return TeamMapper.toResponse(teamRepository.save(teamToSave));
+    }
+
+    public TeamResponse editById(Long id, TeamRequest teamRequest) {
+        return teamRepository.findById(id)
+                .map(teamToEdit -> {
+                    teamToEdit.setOwnerId(teamRequest.ownerId());
+                    teamToEdit.setHackathon(getHackathon(teamRequest.hackathonId()));
+                    teamToEdit.setTeamMembersIds(teamRequest.teamMembersIds());
+                    teamToEdit.setTags(getTags(teamRequest.tags()));
+                    return TeamMapper.toResponse(teamRepository.save(teamToEdit));
+                }).orElseThrow(() -> {
+                    log.error(String.format("Team id = %d not found", id));
+                    return new ResourceNotFoundException(String.format("Team id = %d not found", id));
+                });
+    }
+
+    public TeamResponse editPartialById(Long id, TeamRequest teamRequest) {
+        return teamRepository.findById(id)
+                .map(teamToEdit -> {
+                    if (teamRequest.ownerId() != null) {
+                        teamToEdit.setOwnerId(teamRequest.ownerId());
+                    }
+                    if (teamRequest.hackathonId() != null) {
+                        teamToEdit.setHackathon(getHackathon(teamRequest.hackathonId()));
+                    }
+                    if (teamRequest.teamMembersIds() != null) {
+                        teamToEdit.setTeamMembersIds(teamRequest.teamMembersIds());
+                    }
+                    if (teamRequest.tags() != null) {
+                        teamToEdit.setTags(getTags(teamRequest.tags()));
+                    }
+                    return TeamMapper.toResponse(teamRepository.save(teamToEdit));
+                }).orElseThrow(() -> {
+                    log.error(String.format("Team id = %d not found", id));
+                    return new ResourceNotFoundException(String.format("Team id = %d not found", id));
+                });
+    }
+
+
+    private List<Tag> getTags(List<TagRequest> tags) {
+        val tagsToSave = new ArrayList<Tag>();
+        tags.forEach(tag -> tagsToSave.add(tagService.existsByName(tag.name()) ? tagService.findByName(tag.name()) : tagService.create(tag)));
```

Comment:
wyjaśnij proszę zamysł tego, get który po cichu dorabia tagi, jeśli ich nie znajdzie. Dalej jeśli chodzi o wydajność - dla N tagów strzelasz do bazy 2N razy, można by to mocno uwydajnić a) dodając cache na tagi, b) findByNameIn(names) + create na różnicy między listą z requesta a listą z bazy, najlepiej jeszcze zbatchowane w jeden strzał bo baza będzie płakać